src/app/epics/user-login.epic.spec.ts

```diff
@@ -0,0 +1,176 @@
+import { UserLoginEpic } from './user-login.epic';
+import { ofType, ActionsObservable } from 'redux-observable';
+import { Observable } from 'rxjs/Observable';
+import 'rxjs/add/observable/of';
+import 'rxjs/add/observable/throw';
+
+import 'rxjs/add/observable/empty';
+import 'rxjs/add/operator/switchMap';
+import 'rxjs/add/operator/map';
+import 'rxjs/add/operator/catch';
+
+import { AnyAction } from 'redux';
+import { IUser } from '../models/user.model';
+import { UserActions } from '../actions/user.actions';
+
+let mockUser: IUser = { name: 'MOCK_user' };
+
+class MockLoginService {
+    login = (username, pass) => {};
+
+    logout = () => {};
+};
+// TODO: Don't feel confident in the testing of Observables.
+// TODO: Needs to be updated since Login Service has changed its error handle.
+//      And because there is now a Router dependency
+class MockUserActions {
+    loginSuccess = (user, date) => ({ type: 'MOCK_LOGIN_SUCCESS', user, date });
+    loginFailed = (err) => ({ type: 'MOCK_LOGIN_ERROR', err });
+};
+
+class MockRouter {
+    router = {
+        navigate: (commands, extras) => {},
+    };
+}
+
+
+describe('UserLoginEpic', () => {
+    let mockUserActions: MockUserActions;
+    let mockLoginService: MockLoginService;
+    let mockRouter: MockRouter;
+    let userLoginEpic: UserLoginEpic;
+    let action$: any;
+
+    
+
+    describe('login', () => {
+        beforeEach(() => {
+            mockLoginService = new MockLoginService();
+            mockUserActions = new MockUserActions();
+            mockRouter = new MockRouter();
+            userLoginEpic = new UserLoginEpic(
+                mockLoginService as any, 
+                mockUserActions as any,
+                mockRouter as any
```

Comment:
In this case, when you're declaring your mock such as `let mockUserActions: MockUserActions;`

Rather, type this as `any`
Then when injecting it into `UserLoginEpic`, you can cast it to the appropriate type - this will prevent typescript from complaining, but also afford you hinting and autocomplete `new UserLoginEpic(mockUserActions as MockUserActions, ...)`

Reply:
Wouldn’t it be better to set the initial type as MockService and then cast as the actual Service when injecting? That should give you the autocomplete compared to setting the initial type as ‘any’.